chisel/src/driver.rs

```diff
@@ -0,0 +1,227 @@
+//! Chisel driver implementation.
+//! ChiselDriver is a state machine-like structure implementing the core logic for chisel
+//! execution.
+//! It consumes a ChiselConfig generated by its caller and executes the specified rulesets.
+//! If it enters an error state, it is up to the caller to handle it. If called again, the ruleset
+//! in which the error occurred is dropped.
+//! Upon completed execution, the driver returns a ChiselResult structure.
+
+use std::error::Error;
+use std::fmt::{self, Display};
+use std::fs::{canonicalize, read};
+use std::path::PathBuf;
+
+use libchisel::Module;
+
+use crate::config::ChiselConfig;
+use crate::result::{ChiselResult, ModuleResult, RulesetResult};
+
+/// State machine implementing the main chisel execution loop. Consumes ChiselConfig and returns
+/// ChiselResult, with an intermediate state returned to allow error handling.
+pub struct ChiselDriver {
+    config: ChiselConfig,
+    state: DriverState,
+}
+
+/// The state of the chisel driver.
+pub enum DriverState {
+    Ready,
+    Error(DriverError, ChiselResult),
+    Done(ChiselResult),
+}
+
+pub enum DriverError {
+    /// A module or ruleset is missing a required field. Left-hand is the config object name,
+    /// right-hand is the missing field.
+    MissingRequiredField(String, String),
+    /// The module name was not successfully resolved to an existing chisel module. Left-hand is
+    /// the config object name, right-hand is the invalid module name.
+    ModuleNotFound(String, String),
+    /// A configuration value is of incorrect type or invalid value. Left-hand is the config object
+    /// name, right-hand is the name of the invalid field.
+    InvalidField(String, String),
+    /// A canonicalized path was generated unsuccessfully. Left-hand is the config object name,
+    /// right-hand is the invalid path.
+    PathResolution(String, String),
+    /// An internal error occurred. Field 0 is the config object, during the execution of which the error occurred.
+    /// Field 1 is an additional informational message. Field 2 is the error generated.
+    Internal(String, String, Box<dyn Error>),
+}
+
+impl ChiselDriver {
+    pub fn new(config: ChiselConfig) -> Self {
+        ChiselDriver {
+            config,
+            state: DriverState::Ready,
+        }
+    }
+
+    pub fn take_result(self) -> Result<ChiselResult, ()> {
+        match self.state {
+            DriverState::Ready => Err(()),
+            DriverState::Error(_, result) => Ok(result),
+            DriverState::Done(result) => Ok(result),
+        }
+    }
+
+    pub fn fire(&mut self) -> &DriverState {
```

Comment:
Is there actually a need for this wrapper? It seems the comments from here could be just moved into the match in `fire_internal`.

Reply:
I thought it would be more readable to isolate the guts from the initial matcher, but this was back when we had the huge match statement.

Reply:
Ah that makes sense. Well hope we wont have that monstrosity.